streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -684,19 +684,17 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
                   for {
                     latch <- Promise.make[Nothing, Unit]
                     innerStream = Stream
-                      .bracket(permits.acquire)(_ => permits.release)
-                      .flatMap(_ => ZStream.fromEffect(latch.succeed(())).flatMap(_ => f(a)))
+                      .bracket(permits.acquire *> latch.succeed(()))(_ => permits.release)
```

Comment:
I just realized that acquiring the permits in the bracket is actually unsafe. It makes the acquisition uncancelable.

Reply:
Alright, what do we do now?

Reply:
For a perfect solution, we need a version of `Semaphore#withPermit` that can be lifted to `ZManaged`. It's pretty straightforward to implement.

But for this PR we can just put the permit acquisition back outside the stream. I think it's relatively safe even if not bracketed.

Reply:
Implemented in #1293 and used here. Done.